streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -349,7 +389,7 @@ private static void verifyReceivedAllRecords(final Map<TopicPartition, List<Cons
             final TopicPartition inputTopicPartition = new TopicPartition("data", partitionRecords.getKey().partition());
             final Iterator<ConsumerRecord<byte[], byte[]>> expectedRecord = expectedRecords.get(inputTopicPartition).iterator();
 
-            for (final ConsumerRecord<byte[], byte[]> receivedRecord : partitionRecords.getValue()) {
+            for (final ConsumerRecord<byte[], byte[]> receivedRecord : partitionRecords.getValue().subList(0, partitionRecords.getValue().size() - 1)) {
```

Comment:
Because we do the verification for pending transactions first now, we have one additional record that is not part of the result and that we need to exclude (similar below)

Reply:
The logic looks fragile when the partitionRecords is empty. For all -1 cases, we add one more dummy record to the array being checked, or just remove the last element from the derived array so that we could maintain the same verification.

Reply:
I'm wondering if we have to produce another record in the verifyAllTransactionFinished; for example, could we just check the end value of the newly added `offsets` map maintained by producer? If yes then we can remove this extra logic here and below.

Reply:
> The logic looks fragile when the partitionRecords is empty

`partitionRecords` would never be empty; it would at least contain the "dummy" record that we wrote previously

(We could of course strip the dummy record somewhere else; I picked this solution because it was the least line of code to be changed.)

> for example, could we just check the end value of the newly added offsets map maintained by producer? 

Well, we can only maintain this new offset map _if_ we write those dummy records. If we don't write anything, the producer would not put anything into the map but it would stay empty?

However,  maybe we could use two consumers (and get rid of the producer): one in `read_uncommitted` mode to get the endOffset and a second one with `read_committed` mode that also get the end-offsets in a loop. Only if the "read committed" consumer returns the same end-offset as the "read uncommitted" consumer did, we know that there is no pending transaction?

Thoughts?

Reply:
Yes I was thinking about the added `offsets` map from the `generate()` function -- you're right they would not be shared for the other. Bummer..

I think using two consumer is slightly better than using a producer to write a dummy record.